### PR TITLE
Specify GenerateProposals-9

### PR DESCRIPTION
**Pull request by meiyang-intel**
### Details:
 - *Add spec for GenerateProposals opset9*
 - *GenerateProposals can support paddle generate proposals v2 (https://github.com/PaddlePaddle/PaddleDetection/blob/3c78716d30945e89129c1839453932482f4a6974/ppdet/modeling/ops.py#L1424 and https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/detection/generate_proposals_v2_op.cc) and Caffe2 Generate Proposal(https://github.com/pytorch/pytorch/blob/4dae85c38517fd066a67901f660ca3bc08da7446/caffe2/operators/generate_proposals_op.cc#L365)*
 - *GenerateProposals is a new op based on v6::ExperimentalDetectronGenerateProposalsSingleImage. Below are the changes compared with v6::ExperimentalDetectronGenerateProposalsSingleImage:*
     - Support batch
     - Support dynamic output. Change both outputs to dynamic
     - Add one output 'rpnroisnum'
     - Support one more kind of layout of input 'im_info' which is '[image_height, image_width, scale_height, scale_width]'
     - Swap the execution order of partial sort and ROI size filter
     - Add attribute 'normalized'
     - Add attribute 'nms_eta'

### Tickets:
 - *CVS-79272*
 - *CVS-73120*
 - *CVS-71058*

### CODE REVIEW

**Comment by ceciliapeng2011** [comment]
@meiyang-intel  please refer to liubo's roialign_9 spec pr to fix Documentation/Build_Doc failure. FYI.

**Review by ceciliapeng2011 — APPROVED** [review]

File: docs/ops/detection/GenerateProposals_9.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+The `height` and `width` from inputs `anchors`, `boxesdeltas` and `scores` are the height and width of feature maps.
+
+**Outputs**
+
+* **1**: `rpnrois` - tensor of type *T* with shape `[num_rois, 4]` providing proposed ROIs. The proposals are layouted as `[xmin, ymin, xmax, ymax]`. The `num_rois` means the total proposals number of all the images in one batch.
+
+* **2**: `rpnscores` - tensor of type *T* with shape `[num_rois, 1]` providing proposed ROIs scores.
+
+* **3**: `rpnroisnum` - tensor of type *T_IND* with shape `[num_batches]` providing the number of proposed ROIs in each image.
```
Comment (rkazants) [inline docs/ops/detection/GenerateProposals_9.md:99]: please write as:
tensor of type *roi_num_type* ...

Comment (rkazants) [inline docs/ops/detection/GenerateProposals_9.md:99]: please write as:
tensor of type *roi_num_type* ...

File: docs/ops/detection/GenerateProposals_9.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+**Types**
+
+* *T*: any supported floating-point type.
+* *T_IND*: `int64` or `int32` which is specified from *roi_num_type* values.
```
Comment (rkazants) [inline docs/ops/detection/GenerateProposals_9.md:104]: not needed if you mention *roi_num_type* in `rpnroisnum` description

File: docs/ops/detection/GenerateProposals_9.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+* **4**: `scores` - tensor of type *T* with shape `[num_batches, number_of_anchors, height, width]` providing proposals scores.  **Required.**
+
+The `height` and `width` from inputs `anchors`, `boxesdeltas` and `scores` are the height and width of feature maps.
+
+**Outputs**
+
+* **1**: `rpnrois` - tensor of type *T* with shape `[num_rois, 4]` providing proposed ROIs. The proposals are layouted as `[xmin, ymin, xmax, ymax]`. The `num_rois` means the total proposals number of all the images in one batch.
+
+* **2**: `rpnscores` - tensor of type *T* with shape `[num_rois, 1]` providing proposed ROIs scores.
```
Comment (rkazants) [inline docs/ops/detection/GenerateProposals_9.md:97]: please check why it is not just of shape `[num_rois]`.

Comment (meiyang-intel) [reply docs/ops/detection/GenerateProposals_9.md]: [num_rois] is also ok, have updated.

File: docs/ops/detection/GenerateProposals_9.md (hunk 2 above)
Comment (meiyang-intel) [reply docs/ops/detection/GenerateProposals_9.md]: updated

File: docs/ops/detection/GenerateProposals_9.md (hunk 1 above)
Comment (meiyang-intel) [reply docs/ops/detection/GenerateProposals_9.md]: updated

File: docs/ops/detection/GenerateProposals_9.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    * **Default value**: `i64`
+    * **Required**: *no*
+
+**Inputs**
+
+* **1**: `im_info` - tensor of type *T* and shape `[num_batches, 3]` or `[num_batches, 4]` providing input image info. The image info is layout as `[image_height, image_width, scale_height_and_width]` or as `[image_height, image_width, scale_height, scale_width]`. **Required.**
+
+* **2**: `anchors` - tensor of type *T* with shape `[height, width, number_of_anchors, 4]` providing anchors. Each anchor is layouted as `[xmin, ymin, xmax, ymax]`. **Required.**
+
+* **3**: `boxesdeltas` - tensor of type *T* with shape `[num_batches, number_of_anchors * 4, height, width]` providing deltas for anchors. The delta consists of 4 element tuples with layout `[dx, dy, log(dw), log(dh)]`. **Required.**
```
[the remark this answers is not in the document]
Comment (meiyang-intel) [reply docs/ops/detection/GenerateProposals_9.md:87]: The deltas value are log(dw) and log(dh) in all other frameworks, such as paddle, caffe2, tensorflow. Maybe this is the reason: https://arxiv.org/pdf/1506.01497.pdf in 3.1.2 session.

File: docs/ops/detection/GenerateProposals_9.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+**Detailed description**: The operation performs the following steps for each image:
+
+1.  Transposes and reshapes predicted bounding boxes deltas and scores to get them into the same dimension order as the
+anchors.
+2.  Transforms anchors and deltas into proposal bboxes and clips proposal bboxes to an image. The attribute *normalized*
+indicates whether the proposal bboxes are normalized or not.
+3.  Sorts all `(proposal, score)` pairs by score from highest to lowest; order of pairs with equal scores is undefined.
+4.  Takes top *pre_nms_count* proposals, if total number of proposals is less than *pre_nms_count* takes all proposals.
+5.  Removes predicted boxes with either height or width < *min_size*.
```
Comment (rkazants) [inline docs/ops/detection/GenerateProposals_9.md:18]: In `ExperimentalDetectronGenerateProposalsSingleImage` this step goes at third stage. When it makes sense to do at the third step - any FW requires this? Maybe we need to introduce a separate flag. So we will manage to deprecate `ExperimentalDetectronGenerateProposalsSingleImage`. Please investigate this.

**Review by rkazants — CHANGES_REQUESTED** [review]
Left one comment about possibility to deprecate `ExperimentalDetectronGenerateProposalsSingleImage`

File: docs/ops/detection/GenerateProposals_9.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
+# GenerateProposals {#openvino_docs_ops_detection_GenerateProposals_9}
+
+**Versioned name**: *GenerateProposals-9*
+
+**Category**: *Object detection*
+
+**Short description**: The *GenerateProposals* operation proposes ROIs and their scores
+based on input data for each image in the batch.
```
[the remark this answers is not in the document]
Comment (meiyang-intel) [reply docs/ops/detection/GenerateProposals_9.md:8]: yes, I think it does

File: docs/ops/detection/GenerateProposals_9.md (hunk 5 above)
Comment (meiyang-intel) [reply docs/ops/detection/GenerateProposals_9.md:18]: Do you mean a flag to indicate whether the min_size filter should put before partial sort or after partial sort?
But I see no other framework will put min_size filter before partial sort. They all do min_size filtering after sort like this spec describes.
references:
Caffe2: https://github.com/pytorch/pytorch/blob/4dae85c38517fd066a67901f660ca3bc08da7446/caffe2/operators/generate_proposals_op.cc#L245
Paddle: https://github.com/PaddlePaddle/Paddle/blob/8803f6bb1e390a56c1cb2f0b7dcb754d872f66ab/paddle/fluid/operators/detection/generate_proposals_v2_op.cc#L214
tensorflow: https://github.com/tensorflow/tensorflow/blob/25e649ac3688655547da998eba2715cf70b3e5c9/tensorflow/core/kernels/image/generate_box_proposals_op.cu.cc#L129

I think it may be not necessary to be compatible with ExperimentalDetectronGenerateProposalsSingleImage at this point. What do you think?

**Comment by yuxu42** [comment]
@mitruska could you please review? Thanks!

File: docs/ops/detection/GenerateProposals_9.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+* **4**: `scores` - tensor of type *T* with shape `[num_batches, number_of_anchors, height, width]` providing proposals scores.  **Required.**
+
+The `height` and `width` from inputs `anchors`, `boxesdeltas` and `scores` are the height and width of feature maps.
+
+**Outputs**
+
+* **1**: `rpnrois` - tensor of type *T* with shape `[num_rois, 4]` providing proposed ROIs. The proposals are layouted as `[xmin, ymin, xmax, ymax]`. The `num_rois` means the total proposals number of all the images in one batch.
+
+* **2**: `rpnscores` - tensor of type *T* with shape `[num_rois]` providing proposed ROIs scores.
```
Comment (mitruska) [inline docs/ops/detection/GenerateProposals_9.md:97]: Maybe we can mention that `num_rois` are always dynamic dimension?

**Review by mitruska — APPROVED** [review]

File: docs/ops/detection/GenerateProposals_9.md (hunk 7 above)
Comment (meiyang-intel) [reply docs/ops/detection/GenerateProposals_9.md]: updated
